sky/spot/recovery_strategy.py

```diff
@@ -56,24 +95,52 @@ def make(cls, cluster_name: str, backend: 'backends.Backend',
         # Remove the spot_recovery field from the resources, as the strategy
         # will be handled by the strategy class.
         task.set_resources({resources.copy(spot_recovery=None)})
-        return SPOT_STRATEGIES[spot_recovery](cluster_name, backend, task)
+        return SPOT_STRATEGIES[spot_recovery](cluster_name, backend, task,
+                                              retry_until_up, signal_handler)
 
-    def launch(self,
-               max_retry=3,
-               retry_init_gap_seconds=60,
-               raise_on_failure=True) -> Optional[float]:
+    @_maybe_retry_until_up
+    def launch(self) -> Optional[float]:
         """Launch the spot cluster for the first time.
 
         It can fail if resource is not available. Need to check the cluster
         status, after calling.
 
         Returns: The job's start timestamp, or None if failed to start.
         """
+        return self._launch()
+
+    def recover(self) -> float:
+        """Relaunch the spot cluster after failure and wait until job starts.
+
+        When recover() is called the cluster should be in STOPPED status (i.e.
+        partially down).
+
+        Returns: The timestamp job started.
+        """
+        raise NotImplementedError
+
+    def terminate_cluster(self):
+        """Terminate the spot cluster."""
+        handle = global_user_state.get_handle_from_cluster_name(
+            self.cluster_name)
+        if handle is not None:
+            try:
+                self.backend.teardown(handle, terminate=True)
+            except RuntimeError:
+                logger.error(
+                    f'Failed to terminate the spot cluster {self.cluster_name}.'
+                )
+
+    def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
+        """Implementation of launch()."""
         # TODO(zhwu): handle the failure during `preparing sky runtime`.
         retry_cnt = 0
-        backoff = common_utils.Backoff(retry_init_gap_seconds)
+        backoff = common_utils.Backoff(self.RETRY_INIT_GAP_SECONDS)
```

Comment:
A bit confused that `self.launch()`'s retry already uses retry+backoff, and this inner func also uses retry+backoff. Are both needed? Could we add a comment?